File: Answer/3751.范围内总波动值-i.java
/*
 * @lc app=leetcode.cn id=3751 lang=java
 *
 * [3751] 范围内总波动值 I
 */

// @lc code=start
class Solution {
    private static int[] prefix;

    static {
        final int maxN = 100000;
        int[] waviness = new int[maxN + 1];
        for (int x = 100; x <= maxN; x++) {
            String s = Integer.toString(x);
            int cnt = 0;
            for (int j = 1; j < s.length() - 1; j++) {
                if ((s.charAt(j) > s.charAt(j - 1) && s.charAt(j) > s.charAt(j + 1)) ||
                    (s.charAt(j) < s.charAt(j - 1) && s.charAt(j) < s.charAt(j + 1))) {
                    cnt++;
                }
            }
            waviness[x] = cnt;
        }
        prefix = new int[maxN + 1];
        for (int i = 1; i <= maxN; i++) {
            prefix[i] = prefix[i - 1] + waviness[i];
        }
    }

    public int totalWaviness(int num1, int num2) {
        if (num1 <= 0) {
            return prefix[num2];
        }
        return prefix[num2] - prefix[num1 - 1];
    }
}
// @lc code=end

File: Answer/3751.范围内总波动值-i.py
#
# @lc app=leetcode.cn id=3751 lang=python3
#
# [3751] 范围内总波动值 I
#

# @lc code=start
_MAX_N = 100000
_waviness = [0] * (_MAX_N + 1)
for x in range(100, _MAX_N + 1):
    s = str(x)
    cnt = 0
    for j in range(1, len(s) - 1):
        if (s[j] > s[j - 1] and s[j] > s[j + 1]) or \
           (s[j] < s[j - 1] and s[j] < s[j + 1]):
            cnt += 1
    _waviness[x] = cnt

_prefix = [0] * (_MAX_N + 1)
for i in range(1, _MAX_N + 1):
    _prefix[i] = _prefix[i - 1] + _waviness[i]


class Solution:
    def totalWaviness(self, num1: int, num2: int) -> int:
        if num1 <= 0:
            return _prefix[num2]
        return _prefix[num2] - _prefix[num1 - 1]
# @lc code=end
